---
author: "Wenxuan Jiang"
output: html_document
---

<!-- ## Circos plot -->

<!-- Four required files: -->

<!-- 1. segment data -->
<!--     - define the outmost track, or chromosome position. GRCh36, 37 and 38 are provided. -->
<!--     - three required fields: chromosome name, start position, end position -->

<!-- 2.  -->

<style type="text/css">
  body{
  font-size: 15pt;
}
</style>

## Input format 

This data viewer accepts three types of input formats: .csv, .tsv, and .txt. By default, this viewer reads in the input file as csv. You could adjust the separator used in your variants data set using the `File Extension` drop-down for individual variant type.

This web application provides visualizaitons for three different variant types: Single Nucleotide Variants, Structural Variants, and Copy Number Variants. To take advantage of features in this web application, there are certain required variables for each variant type. 

### Single Nucleotide Variants:

Input file should have the following mandatory fields:

> Sample, Hugo_Symbol, Chromosome, Start_Position, End_Position, Reference_Allele, Tumor_Seq_Allele2, Variant_Classification, Variant_Type, Protein_Change

### Structural Variants:

Input file should have these mandatory fields:

> Sample, Hugo_Symbol, Chromosome1, Start_Position_1, Chromosome2, Start_Position_2

### Copy Number Variants:

Input file should have these mandatory fields:

> Sample, Hugo_Symbol, Chromosome, Start_Position, End_Postion, Reads_Ratio

### Reference Genome:

There are three reference genomes provided for this web application, GRCh38/hg38, GRCh37/hg19, GRCh36/hg18. You can choose the one you used to call variants. By default the dashboard uses GRCh38/hg38. 


**Once you are done uploading your variants datasets and choosing reference genome, you could navigate to User Manual tab to learn how to use this dashboard or go straight to individual variant analysis tab**
